helix-term/src/commands.rs

```diff
@@ -2756,23 +2756,81 @@ fn last_picker(cx: &mut Context) {
 }
 
 // I inserts at the first nonwhitespace character of each line with a selection
+// If the line is empty, automatically indent
 fn insert_at_line_start(cx: &mut Context) {
-    goto_first_nonwhitespace(cx);
-    enter_insert_mode(cx);
+    insert_with_indent(cx, |cursor_line, cursor_line_start, text| {
+        find_first_non_whitespace_char(text.line(cursor_line))
+            .map(|ws_offset| ws_offset + cursor_line_start)
+            .unwrap_or(cursor_line_start)
+    });
 }
 
 // A inserts at the end of each line with a selection
+// If the line is empty, automatically indent
 fn insert_at_line_end(cx: &mut Context) {
+    insert_with_indent(cx, |cursor_line, _, text| {
+        line_end_char_index(text, cursor_line)
+    });
+}
+
+// Enter insert mode and auto-indent the current line if it is empty.
+// If the line is not empty, move the cursor to the specified fallback position.
+fn insert_with_indent(
+    cx: &mut Context,
+    cursor_fallback: impl Fn(usize, usize, &RopeSlice) -> usize,
```

Comment:
Rather than a function that the callers pass in, I think it would be more idiomatic for the callers to pass an enum of `Start`/`Beginning` or `End` and then to have this function do a match:

```rust
let cursor_fallback = match insert_at {
    Start => find_first_non_whitespace_char(text.line(cursor_line)).map(|ws_offset| ws_offset + cursor_line_start).unwrap_or(cursor_line_start),
    End => line_end_char_index(text, cursor_line),
};
```

It would be a little less flexible if `insert_with_indent` were to be used in more places but I don't see other use-cases for that function.

Reply:
A potential use case is if we want to have `xc` insert at indent

Reply:
Agree that this is a bit simpler, I changed it to an enum.